File: app/services/agent_session_parsers/claude_code.rb
module AgentSessionParsers
  class ClaudeCode < Base
    def parse
      records = parse_jsonl_lines
      conversation_records = records.select { |r| r["type"].in?(%w[user assistant]) }

      messages = []
      tool_results_map = build_tool_results_map(conversation_records)

      conversation_records.each do |record|
        msg = record["message"]
        next unless msg

        role = msg["role"]
        timestamp = record["timestamp"]
        raw_content_blocks = msg["content"]

        case role
        when "user"
          content_blocks = parse_user_content(raw_content_blocks)
          next if content_blocks.empty? # Skip tool-result-only messages

          messages << build_message(role: "user", content_blocks: content_blocks, timestamp: timestamp)
        when "assistant"
          content_blocks = parse_assistant_content(raw_content_blocks, tool_results_map)
          next if content_blocks.empty?

          messages << build_message(role: "assistant", content_blocks: content_blocks, timestamp: timestamp)
        end
      end

      metadata = extract_metadata(records, messages)
      build_result(messages: messages, metadata: metadata)
    end

    private

    def build_tool_results_map(records)
      map = {}
      records.each do |record|
        next unless record.dig("message", "role") == "user"

        content = record.dig("message", "content")
        next unless content.is_a?(Array)

        content.each do |block|
          next unless block["type"] == "tool_result"

          tool_use_id = block["tool_use_id"]
          result_content = extract_tool_result_content(block)
          map[tool_use_id] = result_content
        end
      end
      map
    end

    def extract_tool_result_content(block)
      content = block["content"]
      case content
      when String
        truncate_output(content)
      when Array
        content.filter_map { |c| c["text"] if c["type"] == "text" }
          .join("\n")
          .then { |text| truncate_output(text) }
      else
        ""
      end
    end

    def parse_user_content(raw_content)
      case raw_content
      when String
        [text_block(raw_content)]
      when Array
        blocks = []
        raw_content.each do |block|
          case block["type"]
          when "text"
            blocks << text_block(block["text"])
          # Skip tool_result blocks — they're merged into assistant messages
          end
        end
        blocks
      else
        []
      end
    end

    def parse_assistant_content(raw_content, tool_results_map)
      return [] unless raw_content.is_a?(Array)

      blocks = []
      raw_content.each do |block|
        case block["type"]
        when "text"
          blocks << text_block(block["text"])
        when "tool_use"
          result = tool_results_map[block["id"]]
          input_summary = summarize_tool_input(block["name"], block["input"])
          blocks << tool_call_block(
            name: block["name"],
            input: input_summary,
            output: result,
          )
        # Skip "thinking" blocks by default
        end
      end
      blocks
    end

    def summarize_tool_input(name, input)
      return nil unless input.is_a?(Hash)

      case name
      when "Read"
        input["file_path"]
      when "Write"
        input["file_path"]
      when "Edit"
        input["file_path"]
      when "Bash"
        input["command"]
      when "Glob"
        input["pattern"]
      when "Grep"
        "#{input['pattern']} #{input['path']}".strip
      when "Task"
        input["description"] || input["prompt"]&.truncate(100)
      else
        input.to_json.truncate(200)
      end
    end

    def extract_metadata(records, messages)
      first_record = records.find { |r| r["type"].in?(%w[user assistant]) }
      last_record = records.reverse.find { |r| r["type"].in?(%w[user assistant]) }

      {
        "tool_name" => "claude_code",
        "session_id" => first_record&.dig("sessionId"),
        "start_time" => first_record&.dig("timestamp"),
        "end_time" => last_record&.dig("timestamp"),
        "total_messages" => messages.size,
        "working_directory" => first_record&.dig("cwd"),
        "git_branch" => first_record&.dig("gitBranch"),
        "model" => records.find { |r| r.dig("message", "model") }&.dig("message", "model"),
      }.compact
    end
  end
end

File: app/javascript/agentSessionParsers/claudeCode.js
import {
  parseJsonlLines, buildResult, buildMessage, textBlock, toolCallBlock,
  truncateOutput, truncateStr,
} from './base.js';

const CONVERSATION_TYPES = new Set(['user', 'assistant']);

export function parse(rawContent) {
  const records = parseJsonlLines(rawContent);
  const conversationRecords = records.filter(r => CONVERSATION_TYPES.has(r.type));

  const messages = [];
  const toolResultsMap = buildToolResultsMap(conversationRecords);

  for (const record of conversationRecords) {
    const msg = record.message;
    if (!msg) continue;

    const role = msg.role;
    const timestamp = record.timestamp;
    const model = msg.model;
    const rawContentBlocks = msg.content;

    if (role === 'user') {
      const contentBlocks = parseUserContent(rawContentBlocks);
      if (contentBlocks.length === 0) continue;
      messages.push(buildMessage({ role: 'user', contentBlocks, timestamp, model }));
    } else if (role === 'assistant') {
      const contentBlocks = parseAssistantContent(rawContentBlocks, toolResultsMap);
      if (contentBlocks.length === 0) continue;
      messages.push(buildMessage({ role: 'assistant', contentBlocks, timestamp, model }));
    }
  }

  const metadata = extractMetadata(records, messages);
  return buildResult(messages, metadata);
}

function buildToolResultsMap(records) {
  const map = {};
  for (const record of records) {
    if (record.message?.role !== 'user') continue;
    const content = record.message?.content;
    if (!Array.isArray(content)) continue;
    for (const block of content) {
      if (block.type !== 'tool_result') continue;
      const toolUseId = block.tool_use_id;
      const resultContent = extractToolResultContent(block);
      map[toolUseId] = resultContent;
    }
  }
  return map;
}

function extractToolResultContent(block) {
  const content = block.content;
  if (typeof content === 'string') return truncateOutput(content);
  if (Array.isArray(content)) {
    const text = content
      .filter(c => c.type === 'text' && c.text)
      .map(c => c.text)
      .join('\n');
    return truncateOutput(text);
  }
  return '';
}

function parseUserContent(rawContent) {
  if (typeof rawContent === 'string') return [textBlock(rawContent)];
  if (!Array.isArray(rawContent)) return [];

  const blocks = [];
  for (const block of rawContent) {
    if (block.type === 'text') {
      blocks.push(textBlock(block.text));
    }
    // Skip tool_result blocks — they're merged into assistant messages
  }
  return blocks;
}

function parseAssistantContent(rawContent, toolResultsMap) {
  if (!Array.isArray(rawContent)) return [];

  const blocks = [];
  for (const block of rawContent) {
    if (block.type === 'text') {
      blocks.push(textBlock(block.text));
    } else if (block.type === 'tool_use') {
      const result = toolResultsMap[block.id];
      const inputSummary = summarizeToolInput(block.name, block.input);
      blocks.push(toolCallBlock({ name: block.name, input: inputSummary, output: result }));
    }
    // Skip "thinking" blocks by default
  }
  return blocks;
}

function summarizeToolInput(name, input) {
  if (!input || typeof input !== 'object') return undefined;

  switch (name) {
    case 'Read':
    case 'Write':
    case 'Edit':
      return input.file_path;
    case 'Bash':
      return input.command;
    case 'Glob':
      return input.pattern;
    case 'Grep':
      return `${input.pattern || ''} ${input.path || ''}`.trim();
    case 'Task':
      return input.description || truncateStr(input.prompt, 100);
    default:
      return truncateStr(JSON.stringify(input), 200);
  }
}

function extractMetadata(records, messages) {
  const conversationTypes = new Set(['user', 'assistant']);
  const firstRecord = records.find(r => conversationTypes.has(r.type));
  const lastRecord = [...records].reverse().find(r => conversationTypes.has(r.type));

  const meta = {
    tool_name: 'claude_code',
    total_messages: messages.length,
  };
  if (firstRecord?.sessionId) meta.session_id = firstRecord.sessionId;
  if (firstRecord?.timestamp) meta.start_time = firstRecord.timestamp;
  if (lastRecord?.timestamp) meta.end_time = lastRecord.timestamp;
  if (firstRecord?.cwd) meta.working_directory = firstRecord.cwd;
  if (firstRecord?.gitBranch) meta.git_branch = firstRecord.gitBranch;

  const modelRecord = records.find(r => r.message?.model);
  if (modelRecord) meta.model = modelRecord.message.model;

  return meta;
}
